waspc/src/Wasp/Generator/DbGenerator/Jobs.hs

```diff
@@ -27,28 +35,111 @@ migrateDev projectDir = do
   let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
 
+  let job =
+        runNodeCommandAsJob
+          serverDir
+          "npx"
+          [ "--no-install",
+            prismaNpxCommand,
+            "studio",
+            "--schema",
+            SP.toFilePath schemaFile
+          ]
+          J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) jobOutputChannelForwarder
+
+-- | Runs `prisma generate` to generate the Prisma client.
+runGenerate :: Path' Abs (Dir ProjectRootDir) -> J.Job
+runGenerate projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+  let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
+
+  let job =
+        runNodeCommandAsJob
+          serverDir
+          "npx"
+          [ "--no-install",
+            prismaNpxCommand,
+            "generate",
+            "--schema",
+            SP.toFilePath schemaFile
+          ]
+          J.Db
+
+  retryJobOnErrorWith job (npmInstall projectDir) errorOnlyJobOutputChannelForwarder
+
+-- | Runs `npm install` to install dependencies.
+-- May be needed before npx commands if dependencies (e.g. Prisma) are not installed.
+npmInstall :: Path' Abs (Dir ProjectRootDir) -> J.Job
+npmInstall projectDir = do
+  let serverDir = projectDir </> serverRootDirInProjectRootDir
+
   runNodeCommandAsJob
     serverDir
-    "npx"
-    [ "prisma",
-      "studio",
-      "--schema",
-      SP.toFilePath schemaFile
-    ]
+    "npm"
+    ["install"]
     J.Db
+
+-- | Runs the original job, and if there is an error, it attempts recovery by running the second job
+-- before retrying the original one more time.
+-- It also takes a function that handles forwarding channel output.
+retryJobOnErrorWith :: J.Job -> J.Job -> ChannelForwarder -> J.Job
+retryJobOnErrorWith originalJob afterErrorJob channelForwarder externalChannel = do
+  internalChannel <- newChan
+  (initialExitCode, _) <- concurrently (originalJob internalChannel) (channelForwarder internalChannel externalChannel)
+  exitCode <-
+    case initialExitCode of
+      ExitSuccess -> return ExitSuccess
+      ExitFailure _ -> do
+        _ <- concurrently (afterErrorJob internalChannel) (channelForwarder internalChannel externalChannel)
+        (retryExitCode, _) <- concurrently (originalJob internalChannel) (channelForwarder internalChannel externalChannel)
+        return retryExitCode
+  writeJobExitChanMessage exitCode
+  return exitCode
+  where
+    writeJobExitChanMessage exitCode =
+      writeChan
+        externalChannel
+        J.JobMessage
+          { J._data = J.JobExit exitCode,
+            J._jobType = J.Db
+          }
+
+type ChannelForwarder = Chan JobMessage -> Chan JobMessage -> IO ()
+
+jobOutputChannelForwarder :: ChannelForwarder
+jobOutputChannelForwarder fromChan toChan = do
+  message <- readChan fromChan
+  case J._data message of
+    JobOutput _ _ -> writeChan toChan message >> jobOutputChannelForwarder fromChan toChan
+    JobExit _ -> return ()
```

Comment:
From a reply above:

Also, to one of the other comments in here, the reason we have to be careful with returning the `JobExit` is if we allowed the exit to flow from the job on error, and wanted to retry, the actual original caller would stop listening. Here is how these are typically used:

```haskell
  (_, dbSetupResult) <- liftIO (concurrently (readJobMessagesAndPrintThemPrefixed chan) (setupServer genProjectDir chan))
```

So we can only send 1 JobExit message, and that has to be the last thing we send. If we sent the exit from failed job 1, it would not get anything from retry onward.

Reply:
I can at least add comments to help explain this better.

Reply:
Aha, ok this is a good point, thanks for explaining!